### PR TITLE
GLT-1005 Handle non-standard aliases

### PR DESCRIPTION
**Pull request by callunity**
Legacy malformed aliases will need to be maintained, so they and their children can't be run through the alias validators during save.

### CODE REVIEW

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSampleController.java [hunk 1]
```diff
@@ -757,18 +755,14 @@ public ModelAndView setupForm(@PathVariable Long sampleId, @PathVariable Long pr
...
   public ModelAndView editBulkSamples(@PathVariable String sampleIds, ModelMap model) throws IOException {
     try {
-      String[] split = sampleIds.split(",");
-      List<Long> idList = new ArrayList<Long>();
-      for (int i = 0; i < split.length; i++) {
-        idList.add(Long.parseLong(split[i]));
-      }
+      List<Long> idList = getIdsFromString(sampleIds);
       ObjectMapper mapper = new ObjectMapper();
       JSONArray samplesDtos = new JSONArray();
```
Comment (apmasell) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSampleController.java]: Can we do this the other way around and make `sampleDtos` a `List<SampleDto>` and use the Jackson mapper?

File: miso-web/src/main/webapp/pages/editLibrary.jsp [hunk 2]
```diff
@@ -210,6 +210,13 @@
     </c:choose>
 
     <span id="aliasCounter" class="counter"></span>
+    <c:if test="${detailedSample && library.hasNonStandardAlias()}">
+      <ul class="parsley-errors-list filled" id="nonStandardAlias">
+        <li class="parsley-custom-error-message">
+        Double-check thsi alias -- it will be saved even if it is duplicated or malformed!
```
Comment (apmasell) [inline miso-web/src/main/webapp/pages/editLibrary.jsp]: “malformed” is very jargon-y. “does not follow the standard”?

File: miso-service/src/main/resources/log4j.properties [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#
```
Comment (apmasell) [inline miso-service/src/main/resources/log4j.properties]: What's this for?

Comment (callunity) [inline miso-service/src/main/resources/log4j.properties]: I think it's something I added to try to get my test suite working again. Will remove.

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java [hunk 4]
```diff
@@ -813,12 +814,13 @@ public ModelAndView editPropagateSamples(@PathVariable String sampleIds, ModelMa
...
           LibraryAdditionalInfoDto lai = new LibraryAdditionalInfoDto();
+          lai.setNonStandardAlias(((SampleAdditionalInfo) sample).hasNonStandardAlias());
           library.setLibraryAdditionalInfo(lai);
         }
-        libraries.add(library);
+        libraryDtos.add(library);
       }
       model.put("title", "Bulk Create Libraries");
-      model.put("librariesJSON", libraries);
+      model.put("librariesJSON", mapper.writeValueAsString(libraryDtos));
```
Comment (djcooke) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java]: We should add a static utility method somewhere for writing JSON into the model like this, so we don't end up doing it in several different ways again (not necessarily in this PR)

File: sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/sqlstore/SQLLibraryDAO.java [hunk 5]
```diff
@@ -387,7 +387,8 @@ public long save(Library library) throws IOException {
           String name = libraryNamingScheme.generateNameFor("name", library);
           library.setName(name);
           if (libraryNamingScheme.validateField("name", library.getName())
-              && libraryNamingScheme.validateField("alias", library.getAlias())) {
+              && (libraryNamingScheme.validateField("alias", library.getAlias())
+                  || library.getLibraryAdditionalInfo().hasNonStandardAlias())) {
```
Comment (djcooke) [inline sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/sqlstore/SQLLibraryDAO.java]: * Needs null check on LibraryAdditionalInfo
* Same below (L420)
* Above (L380), there is a check for duplicates, which I think should check the nonStandardAlias attribute as well

**Comment by djcooke** [comment]
:+1:

**Comment by apmasell** [comment]
:+1:

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java (hunk 4 above)
Comment (callunity) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java]: I'll make a separate PR for this once I've finished refactoring TagBarcode -> Index

**Pull request merged**